sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/common/security/HeaderBasedAuthenticationMechanism.java

```diff
@@ -70,6 +79,51 @@ public Uni<ChallengeData> getChallenge(RoutingContext context) {
         .item(new ChallengeData(HttpResponseStatus.UNAUTHORIZED.code(), null, null));
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Writes a custom response.
+   */
+  @Override
+  public Uni<Boolean> sendChallenge(RoutingContext context) {
+    // keep the original flow, although I don't need the getChallenge
+    return getChallenge(context)
+        .flatMap(
+            challengeData -> {
+              int status = challengeData.status;
+              context.response().setStatusCode(status);
+
+              // create the response
+              String message =
+                  "Role unauthorized for operation: Missing token, expecting one in the %s header."
+                      .formatted(headerName);
+              ApiError apiError = new ApiError(message, status);
+              try {
+                // try to serialize
+                String response = objectMapper.writeValueAsString(apiError);
```

Comment:
Not sure if it is easy/possible, but serializing as `byte[]` (writeValueAsBytes()) is bit more efficient than writing into intermediate `String`. Fastest would be writing directly into `OutputStream` but that's perhaps not available.

Reply:
Good point.